engine/consensus/matching/engine.go

```
@@ -89,82 +91,44 @@ func NewEngine(
 func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
-	err := e.process(originID, event)
-	if err != nil {
-		if engine.IsIncompatibleInputTypeError(err) {
-			e.log.Warn().Msgf("%v delivered unsupported message %T through %v", originID, event, channel)
-			return nil
-		}
-		return fmt.Errorf("unexpected error while processing engine message: %w", err)
+	receipt, ok := event.(*flow.ExecutionReceipt)
+	if !ok {
+		e.log.Warn().Msgf("%v delivered unsupported message %T through %v", originID, event, channel)
+		return nil
 	}
+	e.addReceiptToQueue(receipt)
 	return nil
 }
 
-// process events for the matching engine on the consensus node.
-func (e *Engine) process(originID flow.Identifier, event interface{}) error {
-	receipt, ok := event.(*flow.ExecutionReceipt)
-	if !ok {
-		return fmt.Errorf("no matching processor for message of type %T from origin %x: %w", event, originID[:],
-			engine.IncompatibleInputTypeError)
-	}
+// addReceiptToQueue adds an execution receipt to the queue of the matching engine, to be processed by a worker
+func (e *Engine) addReceiptToQueue(receipt *flow.ExecutionReceipt) {
 	e.metrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
 	e.pendingReceipts.Push(receipt)
 	e.inboundEventsNotifier.Notify()
-	return nil
 }
 
-// HandleReceipt ingests receipts from the Requester module.
+// HandleReceipt ingests receipts from the Requester module, adding them to the queue.
 func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
 	e.log.Debug().Msg("received receipt from requester engine")
-	err := e.process(originID, receipt)
-	if err != nil {
-		e.log.Fatal().Err(err).Msg("internal error processing event from requester module")
+	r, ok := receipt.(*flow.ExecutionReceipt)
+	if !ok {
+		e.log.Fatal().Err(engine.IncompatibleInputTypeError).Msg("internal error processing event from requester module")
```

I think that ideally HandleReceipt would also be able to use `ctx.Throw` instead of `Log.Fatal`, but the function type is already dictated by being used as a HandleFunc by the Requester engine. Could be pushed to a future refactor of Requester Engine (since that one also still uses `engine.Unit`).

---

I agree that this could be done as part of updating the requester engine. It would be great to have type-safe handler functions in the requester engine, which we could implement by making the requester engine and its Create/Handle functions generic.